setup1-contributor/src/commands/contribute.rs

```diff
@@ -912,33 +909,72 @@ impl Contribute {
     }
 }
 
-fn update_progress_bar(updater: StatusUpdater, progress_bar: ProgressBar) -> tokio::task::JoinHandle<()> {
-    tokio::task::spawn(async move {
-        loop {
-            match updater.status_updater(progress_bar.clone()).await {
-                Ok(_) => {
-                    if progress_bar.is_finished() {
-                        return;
-                    }
-                }
-                Err(e) => {
-                    warn!("Got error from updater: {}", e);
-                    progress_bar.set_message(&format!("Could not update status: {}", e.to_string().trim()));
+async fn update_progress_bar(updater: StatusUpdater) {
+    // This function will only be called if the contributor is already
+    // in the queue. So, we can just print it here and leave it.
+    println!(
+        "You are in the queue for an upcoming round of the ceremony. \
+        Please wait for the prior round to finish, and please stay \
+        connected for the duration of your contribution.",
+    );
+
+    let progress_bar = ProgressBar::new(0);
+    let progress_style =
+        ProgressStyle::default_bar().template("[{elapsed_precise}] {bar:40.cyan/blue} {pos}/{len} {msg}");
+    progress_bar.enable_steady_tick(1000);
+    progress_bar.set_style(progress_style);
+    progress_bar.set_message("Getting initial data from the server...");
+    loop {
+        match updater.status_updater(progress_bar.clone()).await {
+            Ok(_) => {
+                if progress_bar.is_finished() {
+                    return;
                 }
             }
-            sleep(DELAY_POLL_CEREMONY).await;
+            Err(e) => {
+                warn!("Got error from updater: {}", e);
+                progress_bar.set_message(&format!("Could not update status: {}", e.to_string().trim()));
+            }
         }
-    })
+        sleep(DELAY_POLL_CEREMONY).await;
+    }
 }
 
 /// Utility structure to provide updates about the round progress
 struct StatusUpdater {
     server_url: Url,
+    private_key: PrivateKey<Testnet2Parameters>,
     participant_id: String,
 }
 
 impl StatusUpdater {
     async fn status_updater(&self, progress_bar: ProgressBar) -> Result<()> {
+        let status = get_contributor_status(&self.server_url, &self.private_key).await?;
+        match status {
+            ContributorStatus::Queue => {
+                progress_bar.set_message("In the queue...");
+            }
+            ContributorStatus::Round => {
+                self.update_position_in_round(&progress_bar).await?;
+            }
+            ContributorStatus::Finished => {
+                let completed_message = "Successfully contributed, thank you for participation! Waiting to see if you're still needed... Don't turn this off!";
```

Comment:
I am always puzzled by this message. What does it really mean? It is application checking for something?

Reply:
I think it was there initially because it got printed when one of the concurrent tasks would finish - the contributor divides its work over the available threads, so this message would caution the user not to exit yet, in case not all threads were joined. I guess we can update it now though!